Source/WebKit/WebProcess/WebPage/WebPage.cpp

```diff
@@ -6907,13 +6907,20 @@ void WebPage::firstRectForCharacterRangeAsync(const EditingRange& editingRange,
 
     auto range = EditingRange::toRange(*frame, editingRange);
     if (!range)
-        return completionHandler({ }, { });
+        return completionHandler({ }, editingRange);
 
-    // FIXME: Pass an EditingRange that matches the range of the first rect, rather than the entire passed-in range?
     auto rect = RefPtr(frame->view())->contentsToWindow(frame->editor().firstRectForRange(*range));
-    completionHandler(rect, editingRange);
-}
+    auto lineEndPosition = endOfLine(makeContainerOffsetPosition(range->start));
+    auto lineEndBoundary = makeBoundaryPoint(lineEndPosition);
+    if (!lineEndBoundary)
+        return completionHandler({ }, editingRange);
+
+    auto rangeForFirstLine = EditingRange::fromRange(*frame, makeSimpleRange(range->start, WTFMove(lineEndBoundary)));
 
+    rangeForFirstLine.length = std::min(rangeForFirstLine.length, editingRange.length);
```

Comment:
Hm.. I'd expect this to clamp `rangeForFirstLine`, such that its start is at or inside `editingRange`, and the extent is also at or inside `editingRange`. Something kinda like this?

```
rangeForFirstLine.location = std::min(std::max(rangeForFirstLine.location, editingRange.location), editingRange.location + editingRange.length);
rangeForFirstLine.length = std::min(rangeForFirstLine.location + rangeForFirstLine.length, editingRange.location + editingRange.length) - rangeForFirstLine.location;
```